### ISSUE TITLE
stream terminated by RST_STREAM with error code: PROTOCOL_ERROR

### ISSUE THREAD

**Issue by restevez-chs** [grpc/grpc-go#4960]
### What version of gRPC are you using?
v1.42.0

### What version of Go are you using (`go version`)?
go version go1.17.2 darwin/amd64

### What operating system (Linux, Windows, …) and version?
Mac OS 10.15.7

### What did you do?
I have a working API written in Go using grpc and grpc-gateway. When I upgraded to `grpc@v1.42.0`, I expected the API to continue to work.

* Ran `go get -u google.golang.org/grpc@v1.42.0`
* Previous version `google.golang.org/grpc v1.41.0`

### What did you expect to see?
When I upgraded to `grpc@v1.42.0`, I expected the API to continue to work.

### What did you see instead?
I see this error in the response body
```
stream terminated by RST_STREAM with error code: PROTOCOL_ERROR
```

With debug enabled, I see this output in the service logs
```
2021/11/05 17:16:00 ERROR: [transport] transport: http2Server.operateHeaders parsed a :connection header which makes a request malformed as per the HTTP/2 spec
```

From my research, these are the changes causing breakage:
* https://github.com/grpc/grpc-go/pull/4803/files#r715937171
* https://github.com/dfawley/grpc-go/blame/709d464189ac3cced4a52332bbf147e2beeebf61/internal/transport/http2_server.go#L394
* https://github.com/dfawley/grpc-go/commit/689f7b154ee8a3f3ab6a6107ff7ad78189baae06

I have been able to confirm with a Postman request that the error goes away -- and my API request is successful -- if the `Connection` request header is not present. 

However, I am unable to quickly resolve with my Ruby-based Cucumber step definitions where I am using HTTPParty for the API requests. I have also tried net/http. Both of these libraries send a default `Connection` request header, which at the moment I have been unable to remove.

In summary, the stated changes are causing a breakage which is unexpected for a minor version -- the PR [feedback](https://github.com/grpc/grpc-go/pull/4803/files#r715937171) indicates there was a concern during the PR review too. I'm new to Go and GRPC, so any extra feedback would be great. 

Thank you.

**Comment by restevez-chs** [comment]
Update: on my side of things, I was able to resolve using the header filter mechanism with the grpc-gateway. Their [docs](https://grpc-ecosystem.github.io/grpc-gateway/docs/mapping/customizing_your_gateway/#mapping-from-http-request-headers-to-grpc-client-metadata) describe the approach.

TL;DR - the header filter is used to remove the Connection header. 

I still think the breakage in the minor version is questionable, but after more reading, I understand that the change was made to enforce the rule that the Connection header is not to be used with HTTP/2:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection

I look forward to any extra comments and feedback you have. Thank you.

**Comment by dfawley** [comment]
Correct.  This is part of the HTTP/2 spec that we were not following before, so this would be considered a bug fix.  Specifically:

https://httpwg.org/specs/rfc7540.html#rfc.section.8.1.2.2
> An endpoint MUST NOT generate an HTTP/2 message containing connection-specific header fields; any message containing connection-specific header fields MUST be treated as malformed

**Comment by restevez-chs** [comment]
thanks @dfawley
